kong/tools/batch_queue.lua

```diff
@@ -0,0 +1,317 @@
+  if not ok then
+    ngx_log(ERR, "failed to create delayed flush timer: ", err)
+    return
+  end
+  --ngx_log(DEBUG, "delayed timer created")
+  self.flush_scheduled = true
+end
+
+
+-------------------------------------------------------------------------------
+-- Create a timer for the `process` operation.
+-- @param self Queue
+-- @param batch: table with `entries` and `retries` counter
+-- @param delay number: timer delay in seconds
+local function schedule_process(self, batch, delay)
+  local ok, err = timer_at(delay, process, self, batch)
+  if not ok then
+    ngx_log(ERR, "failed to create process timer: ", err)
+    return
+  end
+  self.process_scheduled = true
+end
+
+-----------------
+-- Timer handlers
+-----------------
+
+
+-------------------------------------------------------------------------------
+-- Get the current time.
+-- @return current time in seconds
+local function get_now()
+  return now()*1000
+end
+
+
+-------------------------------------------------------------------------------
+-- Timer callback for triggering a queue flush.
+-- @param premature boolean: ngx.timer premature indicator
+-- @param self Queue
+-- @return nothing
+flush = function(premature, self)
+  if premature then
+    return
+  end
+
+  if get_now() - self.last_t < self.flush_timeout then
+    -- flushing reported: we had activity
+    ngx_log(DEBUG, "[flush] queue had activity, delaying flush")
+    schedule_flush(self)
+    return
+  end
+
+  -- no activity and timeout reached
+  ngx_log(DEBUG, "[flush] queue had no activity, flushing triggered by flush_timeout")
+  self:flush()
+  self.flush_scheduled = false
+end
+
+
+-------------------------------------------------------------------------------
+-- Timer callback for issuing the `self.process` operation
+-- @param premature boolean: ngx.timer premature indicator
+-- @param self Queue
+-- @param batch: table with `entries` and `retries` counter
+-- @return nothing
+process = function(premature, self, batch)
+  if premature then
+    return
+  end
+
+  local next_retry_delay
+
+  local ok, err = self.process(batch.entries)
+  if ok then -- success, set retry_delays to 1
+    self.retry_delay = 1
+    next_retry_delay = 1
+
+  else
+    ngx_log(ERR, "failed to process entries: ", tostring(err))
+
+    batch.retries = batch.retries + 1
+    if batch.retries < self.retry_count then
+      -- queue our data for processing again, at the end of the queue
+      self.batch_queue[#self.batch_queue + 1] = batch
+    else
+      ngx_log(WARN, fmt("entry batch was already tried %d times, dropping it",
+                    batch.retries))
+    end
+
+    self.retry_delay = self.retry_delay + 1
+    next_retry_delay = min(RETRY_MAX_DELAY, self.retry_delay * self.retry_delay)
+  end
+
+  if #self.batch_queue > 0 then -- more to process?
+    ngx_log(DEBUG, fmt("processing oldest data, %d still queued",
+                   #self.batch_queue - 1))
+    local oldest_batch = remove(self.batch_queue, 1)
```

Comment:
remove is not `nil` safe, so when adding entries you should validate `entry ~= nil`